dwave/optimization/src/nodes/mathematical.cpp

```diff
@@ -531,6 +531,272 @@ template class NaryOpNode<functional::min<double>>;
+
+    if constexpr (std::is_same<result_type, bool>::value) {
+        return false;
+    }
+    // there are other cases we could/should handle here.
+    return Array::min();
+}
+
+template <>
+void PartialReduceNode<std::plus<double>>::propagate(State& state) const {
+    auto ptr = data_ptr<PartialReduceNodeData<op>>(state);
+
+    auto& values = ptr->buffer;
+    auto& changes = ptr->updates;
+
+    for (const auto& [p_index, old, value] : array_ptr_->diff(state)) {
+        auto index = _map_parent_index(state, p_index);
+        auto previous = values[index];
+        values[index] += (value - old);
+        changes.emplace_back(index, previous, values[index]);
+    }
+
+    if (ptr->updates.size()) Node::propagate(state);
+}
+
+template <>
+double PartialReduceNode<std::plus<double>>::reduce(const State& state, ssize_t index) const {
+    /// We wish to fill `index` of the partial reduction, given the state of the parent. We proceed
+    /// as follows.
+    /// 1. We unravel the index. This will give us e.g. the indices i, k for
+    ///     R_ik = sum_j P_ijk
+    /// 2. Then we know we have to iterate through the parent starting from j=0 to its full
+    /// dimension. To do so, we first find the starting index in the parent.
+    /// 3. We then iterate through the parent using the strides over the axis we are reducing.
+    assert(index >= 0 && index < this->size(state));
+
+    // 1. Unravel the index we are trying to fill
+    std::vector<ssize_t> indices = unravel_index(this->strides(), index);
+    assert(static_cast<ssize_t>(indices.size()) == this->ndim());
+
+    // 2. Find the respective starting index in the parent
+    ssize_t start_idx = 0;
+    for (ssize_t ax = 0; ax < this->ndim(); ++ax) {
+        if (ax >= this->axis_) {
+            start_idx += indices[ax] * parent_strides_c_[ax + 1] / array_ptr_->itemsize();
+        } else {
+            start_idx += indices[ax] * parent_strides_c_[ax] / array_ptr_->itemsize();
+        }
+    }
+
+    ssize_t begin = 0;
+
+    // get the initial value
+    double init;
+    if (this->init.has_value()) {
+        init = this->init.value();
+    } else {
+        // if there is no init, we use the first value in the array as the init
+        // we should only be here if the array is not dynamic and if it has at
+        // least one entry
+        assert(!array_ptr_->dynamic());
+        assert(array_ptr_->size(state) >= 1);
+        init = array_ptr_->view(state)[start_idx + begin];
+        ++begin;
+    }
+
+    /// 3. Iterate through the parent on the correct axis.
+    double val = init;
+    auto func = op();
+    for (ssize_t i = begin; i < array_ptr_->shape()[this->axis_]; ++i) {
+        ssize_t p_index =
+                start_idx + i * this->parent_strides_c_[this->axis_] / array_ptr_->itemsize();
+        assert(p_index < static_cast<ssize_t>(array_ptr_->size(state)));
+        double other = array_ptr_->view(state)[p_index];
```

Comment:
It feels a bit redundant to keep the strides as if the parent was contiguous, then also use `view` to account for the parent's strides.

Maybe worth just working with the parent's actual strides directly?

Reply:
Good point, let me rethink about the logic. I'll reply here later on.

Reply:
so here you are saying something like
```
ssize_t p_index =
                start_idx + i *array_ptr_->strides()[this->axis_] / array_ptr_->itemsize();
        assert(p_index < static_cast<ssize_t>(array_ptr_->size(state)));
        double other = array_ptr_->buff(state)[p_index];
```
if I understand correctly?

Reply:
I think I have a snippet. Putting it together now.

Reply:
Update this now, let me know if it looks good